### PR TITLE
Feature/port and adapter api

### PR DESCRIPTION
**Pull request by bcncpp**
Initial Design of the Inference Engine with the models directory structure

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -14,4 +14,8 @@ target/
 #  be found at https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore
 #  and can be added to the global gitignore or merged into this file.  For a more nuclear
 #  option (not recommended) you can uncomment the following to ignore the entire idea folder.
-#.idea/
\ No newline at end of file
+.idea/
+
```
Comment (bcncpp) [inline .gitignore:18]: do you use jetbrains rover?

File: src/galemind/api/inference/mod.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+pub enum InferenceResponse {
+    Ok(Inference),
+    Error(InferenceError)
+}
+
+pub struct Inference {
+    model_name: String,
+    model_version: Option<String>,
```
Comment (bcncpp) [inline src/galemind/api/inference/mod.rs:38]: We might want to configure the backend dynamically for each request, assign TPU/GPU for each training. We need to investigate in triton how it does.

File: src/galemind/api/inference/mod.rs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    parameters: Option<HashMap<String, InferParameter>>,
+    data: tensor::Data
+}
+
+pub struct InferenceError {
+    error: String
+}
+
+pub trait InferenceProcessor {
+    fn process(&self, request: InferenceRequest) -> InferenceResponse;
```
Comment (bcncpp) [inline src/galemind/api/inference/mod.rs:57]: The inference processor can have a backend. We aim to support ONNX, Pytorch, Tensor Flow probally we need to decouple as in my PR.
``graph TD
  A[Client Request] --> B[API Gateway / Load Balancer]
  B --> C[Authentication & Authorization]
  C --> D[Preprocessing Layer]
  D --> E[Model Inference Engine]
  E --> F[Model Server (e.g., TensorFlow Serving, TorchServe)]
  F --> G[ML Model]
  E --> H[Postprocessing Layer]
  H --> I[Response Formatting]
  I --> J[Send Response to Client]

  subgraph "Model Server Infrastructure"
    F
    G
  end`

File: .gitignore (hunk 1 above)
Comment (alexandre-cremieux) [reply .gitignore:18]: yes

File: src/galemind/api/inference/mod.rs (hunk 2 above)
Comment (alexandre-cremieux) [reply src/galemind/api/inference/mod.rs:38]: In fact, this model is the one that will exposed to the outside world. It is not intended to expose the inner logic of an inference service.

But yes, I had the same reflection that we will need other parameters to run an engine. We will probably need a mediator between the inference service and the targeted engine.

File: src/galemind/api/inference/mod.rs (hunk 3 above)
Comment (bcncpp) [reply src/galemind/api/inference/mod.rs]: Let's make this proposal working for the dummy case.

**Comment by giorgiozoppi** [comment]
I added a train script and fixed the CI and the devcontainer. I have also added tests for the fake processor and a makefile. I leave to you to continue the missing design parts.

**Comment by giorgiozoppi** [comment]
Continue modelling. The work is quite good, think also to the async case. Waiting is hell.

File: src/galemind/api/inference/mod.rs (hunk 3 above)
Comment (bcncpp) [reply src/galemind/api/inference/mod.rs]: We might also need a request queue if we take in account a scheduler.

File: models/src/gm_models/model/__pycache__/__init__.cpython-311.pyc
Comment (bcncpp) [inline models/src/gm_models/model/__pycache__/__init__.cpython-311.pyc:1]: this should removed and put .gitgnore

File: src/galemind/api/inference/mod.rs
Comment (markojeremic) [inline src/galemind/api/inference/mod.rs:1]: From my starting point of view, I think that we should use files named 'mod.rs' for including other modules at the first. As this code logic for inference will grow for sure, it will be hard to track these changes for inference if they stay in one single 'mod.rs' file as their primary usage is to include other modules as I mentioned above. It is much clearer for me if we would separate this file into two files for now:
1. inference_processor.rs which will have this logic below,
2. mod.rs file which will include and use inference_processor module.

So, for example, if we want to add new logic for inference errors in the future, we would follow the flow:
1. create inference_errors.rs file and add logic for handling inference errors,
2. include and use that inference errors module in mod.rs file

File: src/galemind/api/model/mod.rs
Comment (markojeremic) [inline src/galemind/api/model/mod.rs:1]: Same thoughts and proposals as for inference/mod.rs file 👍

File: src/galemind/api/tensor.rs
Comment (markojeremic) [inline src/galemind/api/tensor.rs:1]: Should we keep tensor.rs file directly in api root directory? It makes more sense for me to be placed in some directory.

File: src/galemind/inference/mod.rs
Comment (markojeremic) [inline src/galemind/inference/mod.rs:1]: I like this approach (having separate mod.rs and fake.rs files for inference). It is the same as I proposed for api/inference/mod.rs and api/model/mod.rs files. We should try to be consistent as much as possible and try to stick to some common conventions.

File: src/galemind/api/inference/mod.rs
Comment (bcncpp) [reply src/galemind/api/inference/mod.rs:1]: Do you want error separated to another file or other module?

File: README.md [hunk 4]
```diff
@@ -13,6 +13,21 @@
 
 ---
 
+## 🗂 Repository Structure
+
+```
+galemind/
+├── .devcontainer/     # Configuration for development containers
+├── .github/           # GitHub action workflows
+├── engine/            # Core engine logic for the inference server
```
[the remark this answers is not in the document]
Comment (bcncpp) [reply README.md:22]: Yes. Agree. We keep consistency.

File: .github/workflows/ci.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        uses: actions/checkout@v4
+
+      - name: Set up Rust
+        uses: actions-rs/toolchain@v1
+        with:
+            toolchain: stable
+            override: true
+            components: rustfmt, clippy
+
+      - name: Cache cargo registry
```
Comment (bcncpp) [inline .github/workflows/ci.yaml:26]: We can apply also clippy.
